doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,359 @@
+recover the original validity period. We can define the three parts like so:
+
+1. `Q_safemfp` will contain the non-temporal part of `Q_mfp` combined with two additional `Map`
+   expressions. These two additional map expressions will correspond to the lower and upper bounds
+   and will be calculating the final timestamp for them. We can do so by fabricating two
+   `MirScalarExpr`s. One that calculates the maximum of all the lower bound exprs and one that
+   calculates the minimum of all the upper bound exprs. Effectively, we will precompute the validity
+   period and store it as two additional timestamp columns at the end of the original row.
+2. `Q_expiration` will contain the temporal predicate `mz_now() <= upper_ts` which ensures each
+   record is retracted at the upper bound of its validity period.
+3. `Q_final` will contain the temporal predicate `lower_ts <= mz_now()` and a projection that hides
+   the last two columns, which contain the two timestamps of the validity period. This will ensure
+   that records are only inserted at the lower bound of their validity period and that `Q_expr` will
+   see rows with the correct schema.
+
+Given this decomposition we can now define the ingestion as:
+
+```
+C = Q_final(PERSIST(Q_expiration(Q_safemfp(RAW))))
+```
+
+Note that in the equation above `Q_final` is applied on the *output* of durably persisting
+persisting the rest of the ingestion. This MFP fragment will be deferred and applied on the read
+side when a dataflow uses the collection `C`.
+
+With these equations we have accomplished the goal of doing as much work as possible before the
+first persist operator while requiring no past state from `RAW` to resume. To see why that is the
+case, let's describe the state needed by the `Q_expiration` operator. The state of `Q_expiration` at
+time `t` is the set of records `(row, t_record, diff)` for which `t_record <= t && t <
+upper_ts(Q_mfp(row))`. In other words, for every timestamp it needs to remember all past records
+that expire after that timestamp.
+
+If we now look at the output of `Q_safemfp` we see that it converts a `(row, t_record, diff)` update
+to `((row, lower_ts, upper_ts), t_record, diff)` update. Additionally, `Q_expiration` will transform
+each `((row, lower_ts, upper_ts), t_record, diff)` update into two:
+
+1. `((row, lower_ts, upper_ts), t_record, diff)`
+2. `((row, lower_ts, upper_ts), upper_ts, -diff)`
+
+Given the above we can see that at the beginning of an ingestion execution `Q_expiration` can
+recover its in-memory state at time `t` by taking a snapshot of `C_intermediate` (not `RAW`) at time
+`t`, which is available because it is stored in persist! This is because any record that may expire
+in the future is maintained until the end of its validity period and `Q_mfp` has preserved the
+previously computed timestamps. This technique is similar to how the `UPSERT` envelope recovers its
+state from its output.
```

Comment:
Just flagging that I largely skipped this section due to bandwidth. Made sense on a skim, but I didn't actually think through the equations in depth.